### PR TITLE
Set expire after we sent close in ZooKeeper client

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug when clickhouse-server doesn't send `close` request to ZooKeeper server.

### CODE REVIEW

**Comment by alesapin** [comment]
Just noticed that while implementing TestKeeperStorage. Seems like we never sent `close` request to the ZooKeeper server during normal shutdown.

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 1]
```diff
@@ -1342,34 +1343,39 @@ void ZooKeeper::receiveEvent()
...
 void ZooKeeper::finalize(bool error_send, bool error_receive)
 {
-    {
-        std::lock_guard lock(push_request_mutex);
-
-        if (expired)
-            return;
-        expired = true;
-    }
+    if (expired)
```
Comment (nikitamikhaylov) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1346]: Possible race with line 1365 and line 1376 because of the absence of lock_guard ?

**Comment by alesapin** [comment]
Hmm, what's wrong with `test_distributed_ddl` ?

File: src/Common/ZooKeeper/ZooKeeperImpl.cpp [hunk 2]
```diff
@@ -1342,34 +1343,45 @@ void ZooKeeper::receiveEvent()
 
 void ZooKeeper::finalize(bool error_send, bool error_receive)
 {
+    bool check = false;
+    /// If some thread (send/receive) already finalizing session don't try to do it
+    if (!finalization_started.compare_exchange_strong(check, true))
```
Comment (nikitamikhaylov) [inline src/Common/ZooKeeper/ZooKeeperImpl.cpp:1348]: I think `exchange` will be enough, because it returns previous value.

Comment (alesapin) [reply src/Common/ZooKeeper/ZooKeeperImpl.cpp:1348]: Much simplier, thanks

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by alesapin** [comment]
Tests Ok.

**Pull request merged**